interfaces/cython/cantera/test/test_thermo.py

```diff
@@ -1520,3 +1521,156 @@ def test_write_csv(self):
         self.assertEqual(len(data), 7)
         self.assertEqual(len(data.dtype), self.gas.n_species + 2)
         self.assertIn('Y_H2', data.dtype.fields)
+
+        b = ct.SolutionArray(self.gas)
+        b.read_csv(outfile)
+        self.assertTrue(np.allclose(states.T, b.T))
+        self.assertTrue(np.allclose(states.P, b.P))
+        self.assertTrue(np.allclose(states.X, b.X))
+
+    def test_pandas(self):
+        states = ct.SolutionArray(self.gas, 7)
+        states.TPX = np.linspace(300, 1000, 7), 2e5, 'H2:0.5, O2:0.4'
+        try:
+            # this will run through if pandas is installed
+            df = states.to_pandas()
+            self.assertTrue(df.shape[0]==7)
+        except ImportError as err:
+            # pandas is not installed and correct exception is raised
+            pass
+        except Exception as err:
+            raise(err)
+
+    def test_restore(self):
+
+        def check(a, b, atol=None):
+            if atol is None:
+                fcn = lambda c, d: np.allclose(c, d)
+            else:
+                fcn = lambda c, d: np.allclose(c, d, atol=atol)
+            check = fcn(a.T, b.T)
+            check &= fcn(a.P, b.P)
+            check &= fcn(a.X, b.X)
+            return check
+
+        # test ThermoPhase
+        a = ct.SolutionArray(self.gas)
+        for i in range(10):
+            T = 300 + 1800*np.random.random()
+            P = ct.one_atm*(1 + 10*np.random.random())
+            X = np.random.random(self.gas.n_species)
+            X[-1] = 0.
+            X /= X.sum()
+            a.append(T=T, P=P, X=X)
+
+        data, labels = a.collect_data()
+
+        # basic restore
+        b = ct.SolutionArray(self.gas)
+        b.restore_data(data, labels)
+        self.assertTrue(check(a, b))
```

Comment:
Rather than using `assertTrue(check(...))`, I would recommend making `check` a member function which calls other assert functions. This will help provide more granular error messages, e.g. you will know exactly which of the arrays didn't match.

Reply:
Excellent suggestion.